libdrgn/arch_ppc64.c.in

```diff
@@ -0,0 +1,323 @@
+	dwarf_regs[23] = READ_REGISTER(23);
+	dwarf_regs[24] = READ_REGISTER(24);
+	dwarf_regs[25] = READ_REGISTER(25);
+	dwarf_regs[26] = READ_REGISTER(26);
+	dwarf_regs[27] = READ_REGISTER(27);
+	dwarf_regs[28] = READ_REGISTER(28);
+	dwarf_regs[29] = READ_REGISTER(29);
+	dwarf_regs[30] = READ_REGISTER(30);
+	dwarf_regs[31] = READ_REGISTER(31);
+
+	dwarf_ctr = READ_REGISTER(35);
+	dwarf_link = READ_REGISTER(36);
+	dwarf_xer = READ_REGISTER(37);
+#undef READ_REGISTER
+
+	if (!dwfl_thread_state_registers(thread, 0, 32, dwarf_regs))
+		return drgn_error_libdwfl();
+
+	if (!dwfl_thread_state_registers(thread, 65, 1, &dwarf_link))
+		return drgn_error_libdwfl();
+
+	if (!dwfl_thread_state_registers(thread, 101, 1, &dwarf_ctr))
+		return drgn_error_libdwfl();
+
+	if (!dwfl_thread_state_registers(thread, 109, 1, &dwarf_xer))
+		return drgn_error_libdwfl();
+
+	return NULL;
+}
+
+static struct drgn_error *
+pt_regs_set_initial_registers_ppc64(Dwfl_Thread *thread,
+		                     const struct drgn_object *obj)
+{
+	bool bswap = (obj->value.little_endian !=
+		      (__BYTE_ORDER__ == __ORDER_LITTLE_ENDIAN__));
+	return set_initial_registers_from_struct_ppc64(thread,
+						       drgn_object_buffer(obj),
+						       drgn_buffer_object_size(obj),
+						       bswap);
+}
+
+static struct drgn_error *
+prstatus_set_initial_registers_ppc64(struct drgn_program *prog,
+				     Dwfl_Thread *thread, const void *prstatus,
+				     size_t size)
+{
+	if (size < 112) {
+		return drgn_error_create(DRGN_ERROR_INVALID_ARGUMENT,
+				"NT_PRSTATUS is truncated");
+	}
+	bool bswap;
+        struct drgn_error *err = drgn_program_bswap(prog, &bswap);
+	if (err)
+                return err;
+        return set_initial_registers_from_struct_ppc64(thread,
+                                                        (char *)prstatus + 112,
+                                                        size - 112, bswap);
+}
+
+static inline struct drgn_error *read_register(struct drgn_object *reg_obj,
+		struct drgn_object *frame_obj,
+		const char *name,
+		Dwarf_Addr *ret)
+{
+	struct drgn_error *err;
+	uint64_t reg;
+
+	err = drgn_object_member_dereference(reg_obj, frame_obj, name);
+	if (err)
+		return err;
+
+	err = drgn_object_read_unsigned(reg_obj, &reg);
+	if (err)
+		return err;
+	*ret = reg;
+
+	return NULL;
+}
+
+#define LR_OFFSET	16
+
+static struct drgn_error *
+linux_kernel_set_initial_registers_ppc64(Dwfl_Thread *thread,
+					 const struct drgn_object *task_obj)
+{
+	struct drgn_error *err;
+	struct drgn_program *prog = drgn_object_program(task_obj);
+	struct drgn_object sp_obj, pt_regs_obj, reg_obj;
+	union drgn_value value_mem;
+	const union drgn_value *value;
+	Dwarf_Word dwarf_reg;
+
+	drgn_object_init(&sp_obj, prog);
+	drgn_object_init(&pt_regs_obj, prog);
+	drgn_object_init(&reg_obj, prog);
+
+	err = drgn_object_member_dereference(&pt_regs_obj, task_obj, "thread");
+	if (err)
+		goto out;
+
+	err = drgn_object_member(&pt_regs_obj, &pt_regs_obj, "regs");
```

Comment:
No worries, thanks for the review.  I guess, it can be done, by passing the `task_struct->thread_struct->pt_regs` to  `pt_regs_set_initial_registers_ppc64()` in `linux_kernel_set_initial_registers_ppc64()`, also the code to adjust the stack pointer to skip the first frame (unreliable), needs to stay along with link register (return address) adjustment code based on the new frame.

Reply:
Something like https://gist.github.com/kamalesh-babulal/991f904246ff1975232ffb4aa02ccd80

Reply:
A follow-up question since I'm not familiar with ppc: what about the first frame is unreliable?

Reply:
AFAIK, The reason for bottom-most stack frame being unreliable for the scheduled out task, is that the frame is completely uninitialized and if the task was in do_IRQ() (interrupt), it will have residual of exception marker in the first frame but in the both cases complying to the "Power Architecture 64-Bit ELF V2 ABI", the Link Register (LR) will contain the back chain pointing to the previous frame.